File: picasso-commons/tests/src/com/hendraanggrian/picasso/commons/test/InstrumentedTest.java
package com.hendraanggrian.picasso.commons.test;

import android.os.Build;
import android.os.CountDownTimer;
import android.support.annotation.NonNull;
import android.support.test.espresso.UiController;
import android.support.test.espresso.ViewAction;
import android.support.test.rule.ActivityTestRule;
import android.support.test.runner.AndroidJUnit4;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;

import com.hendraanggrian.picasso.commons.target.Targets;
import com.hendraanggrian.picasso.commons.transformation.Transformations;
import com.squareup.picasso.Picasso;

import org.hamcrest.Matcher;
import org.junit.FixMethodOrder;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.MethodSorters;

import static android.support.test.InstrumentationRegistry.getTargetContext;
import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.matcher.ViewMatchers.isAssignableFrom;
import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
import static android.support.test.espresso.matcher.ViewMatchers.withId;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
@RunWith(AndroidJUnit4.class)
@FixMethodOrder(MethodSorters.NAME_ASCENDING)
public class InstrumentedTest {

    private static final long DELAY_COUNTDOWN = 3000;

    @Rule
    public ActivityTestRule<InstrumentedActivity> rule = new ActivityTestRule<>(InstrumentedActivity.class);

    @Test
    public void test1_transformation() {
        onView(withId(R.id.imageView))
                .perform(
                        new ViewAction() {
                            @Override
                            public Matcher<View> getConstraints() {
                                return isAssignableFrom(ImageView.class);
                            }

                            @Override
                            public String getDescription() {
                                return "test1_transformation()";
                            }

                            @Override
                            public void perform(UiController uiController, View view) {
                                Picasso.with(getTargetContext())
                                        .load(R.drawable.bg_test)
                                        .transform(Transformations.circle())
                                        .into((ImageView) view);
                            }
                        },
                        delay());
    }

    @Test
    public void test2_placeholder() {
        onView(withId(R.id.imageView))
                .perform(
                        new ViewAction() {
                            @Override
                            public Matcher<View> getConstraints() {
                                return isAssignableFrom(ImageView.class);
                            }

                            @Override
                            public String getDescription() {
                                return "test2_placeholder()";
                            }

                            @Override
                            public void perform(UiController uiController, View view) {
                                Picasso.with(getTargetContext())
                                        .load("https://i.ytimg.com/vi/yaqe1qesQ8c/maxresdefault.jpg")
                                        .into(Targets.placeholder((ImageView) view));
                            }
                        },
                        delay());
    }

    @NonNull
    public ViewAction delay() {
        return new ViewAction() {
            @Override
            public Matcher<View> getConstraints() {
                return isDisplayed();
            }

            @Override
            public String getDescription() {
                return "delay for " + DELAY_COUNTDOWN;
            }

            @Override
            public void perform(UiController uiController, View view) {
                final ProgressBar progressBar = rule.getActivity().progressBar;
                new CountDownTimer(DELAY_COUNTDOWN, 100) {
                    @Override
                    public void onTick(long millisUntilFinished) {
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                            progressBar.setProgress((int) (progressBar.getMax() * millisUntilFinished / DELAY_COUNTDOWN), true);
                        } else {
                            progressBar.setProgress((int) (progressBar.getMax() * millisUntilFinished / DELAY_COUNTDOWN));
                        }
                    }

                    @Override
                    public void onFinish() {
                        progressBar.setVisibility(View.GONE);
                    }
                }.start();
                uiController.loopMainThreadForAtLeast(DELAY_COUNTDOWN);
            }
        };
    }
}

File: picasso-utils/tests/src/com/hendraanggrian/picasso/test/InstrumentedTest.kt
package com.hendraanggrian.picasso.test

import android.os.Build
import android.os.CountDownTimer
import android.support.test.InstrumentationRegistry.getTargetContext
import android.support.test.espresso.Espresso.onView
import android.support.test.espresso.UiController
import android.support.test.espresso.ViewAction
import android.support.test.espresso.matcher.ViewMatchers.*
import android.support.test.rule.ActivityTestRule
import android.support.test.runner.AndroidJUnit4
import android.view.View
import android.widget.ImageView
import com.hendraanggrian.picasso.picasso
import com.hendraanggrian.picasso.target.Targets
import com.hendraanggrian.picasso.transformation.Transformations
import org.hamcrest.Matcher
import org.junit.FixMethodOrder
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.MethodSorters

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
@RunWith(AndroidJUnit4::class)
@FixMethodOrder(MethodSorters.NAME_ASCENDING)
class InstrumentedTest {

    companion object {
        private val DELAY_COUNTDOWN = 3000L
    }

    @Rule @JvmField var rule = ActivityTestRule(InstrumentedActivity::class.java)

    @Test
    fun test1_transformation() {
        onView(withId(R.id.imageView))
                .perform(
                        object : ViewAction {
                            override fun getConstraints(): Matcher<View> = isAssignableFrom(ImageView::class.java)
                            override fun getDescription(): String = "test1_transformation()"
                            override fun perform(uiController: UiController, view: View) = getTargetContext()
                                    .picasso(R.drawable.bg_test)
                                    .transform(Transformations.circle())
                                    .into(view as ImageView)
                        },
                        delay())
    }

    @Test
    fun test2_placeholder() {
        onView(withId(R.id.imageView))
                .perform(
                        object : ViewAction {
                            override fun getConstraints(): Matcher<View> = isAssignableFrom(ImageView::class.java)
                            override fun getDescription(): String = "test2_placeholder()"
                            override fun perform(uiController: UiController, view: View) = getTargetContext()
                                    .picasso("https://i.ytimg.com/vi/yaqe1qesQ8c/maxresdefault.jpg")
                                    .into(Targets.placeholder(view as ImageView))
                        },
                        delay())
    }

    fun delay(): ViewAction = object : ViewAction {
        override fun getConstraints(): Matcher<View> = isDisplayed()
        override fun getDescription(): String = "delay for " + DELAY_COUNTDOWN
        override fun perform(uiController: UiController, view: View) {
            val progressBar = rule.activity.progressBar!!
            object : CountDownTimer(DELAY_COUNTDOWN, 100) {
                override fun onTick(millisUntilFinished: Long) {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                        progressBar.setProgress((progressBar.max * millisUntilFinished / DELAY_COUNTDOWN).toInt(), true)
                    } else {
                        progressBar.progress = (progressBar.max * millisUntilFinished / DELAY_COUNTDOWN).toInt()
                    }
                }

                override fun onFinish() {
                    progressBar.visibility = View.GONE
                }
            }.start()
            uiController.loopMainThreadForAtLeast(DELAY_COUNTDOWN)
        }
    }
}